rfcs/convergence/as-prop.md

```diff
@@ -0,0 +1,197 @@
+# RFC: Simplify handling of `as` prop
+
+---
+
+@layershifter @levithomason
+
+## Summary
+
+Restrict `as` prop to accept only HTML elements due issues with typings and problems with developer experience.
+
+## Typings problem
+
+Fluent UI components should support `as` prop to allow consumers to customize HTML tags, it's especially useful for components like `Text`:
+
+```tsx
+<>
+  {/* renders <span /> */}
+  <Text />
+  {/* renders <p /> */}
+  <Text as="p" />
+</>
+```
+
+However, currently components are accepting the whole `React.ElementType` in [`as` prop](https://github.com/microsoft/fluentui/blob/7f5086718eec496063c6830302c162117fcfc4ec/packages/react-utilities/src/compose/types.ts#L15):
+
+```tsx
+<Button as={Link} /> // renders <Link /> + additional props
+```
+
+These usages are nightmare for typings as we must use really complicated generics ([`ComponentWithAs`](https://github.com/microsoft/fluentui/blob/c27c8fa7ae589a61267bd0a6f0f1c7a49a6be9c2/packages/react-compose/src/types.ts#L21)) to support polymorphism. There are few similar implementations:
+
+- [react-polymorphic-box](https://github.com/kripod/react-polymorphic-box)
+- [forwardRefWithAs from Reach UI](https://github.com/reach/reach-ui/blob/e1a6fba7042e1e023c4a3a4e7c15a13225500667/packages/utils/src/index.tsx#L191-L199)
+
+Anyway, with these generics our component can still throw on unsupported props:
+
+```tsx
+// ✅ TS compiler will throw an error because "someProp" is not supported by "Button" and "Link"
+<Button as={Link} someProp />
+```
+
+But, it still creates problems for customers when they will decide to compose our components:
+
+```tsx
+// Uses a React HOC to control rendering
+// https://reactjs.org/docs/react-api.html#reactmemo
+const MemoButton = React.memo(Button);
+
+// ⚠ It should be enough for customers to use "ButtonProps", but "ComponentWithAs" requires different typings
+function PrimaryButton(props: ButtonProps) {
+  return <Button {...props} primary />;
+}
+
+function App() {
+  return (
+    <>
+      {/* ✅ TS throws as "href" is not supported for "div" elements */}
+      <Button as="div" href="#" />
+      {/* no errors 🤷‍♂️ */}
+      <MemoButton as="div" href="#" />
+
+      {/* 💣 "as" and appropriate props are created by "ComponentWithAs" generic */}
+      {/*     generic thus are not inside "ButtonProps" */}
+      <CustomButton as="div" />
```

Comment:
Confused here, what is `CustomButton` in this example ?

Reply:
Oops, it should be "PrimaryButton", fixed 👍